pkgs/cli_config/lib/src/config.dart

```diff
@@ -0,0 +1,387 @@
+      _cliProvider: splitCliPattern,
+      _environmentProvider: splitEnvironmentPattern,
+      _fileProvider: null
+    }.entries) {
+      final provider = entry.key;
+      final splitPattern = entry.value;
+      final value =
+          provider.getOptionalStringList(key, splitPattern: splitPattern);
+      if (value != null) {
+        if (combineAllConfigs) {
+          (result ??= []).addAll(value);
+        } else {
+          return value;
+        }
+      }
+    }
+    return result;
+  }
+
+  static const boolStrings = {
+    '0': false,
+    '1': true,
+    'false': false,
+    'FALSE': false,
+    'true': true,
+    'TRUE': true,
+  };
+
+  /// Lookup a boolean value in this config.
+  ///
+  /// First tries CLI argument defines, then environment variables, and
+  /// finally the config file.
+  ///
+  /// For cli defines and environment variables, the value must be one of
+  /// [boolStrings].
+  /// For the config file, it must be a boolean.
+  ///
+  /// Throws if one of the configs does not contain the expected value type.
+  bool getBool(String key) {
+    final value = getOptionalBool(key);
+    _throwIfNull(key, value);
+    return value!;
+  }
+
+  /// Lookup an optional boolean value in this config.
+  ///
+  /// First tries CLI argument defines, then environment variables, and
+  /// finally the config file.
+  ///
+  /// For cli defines and environment variables, the value must be one of
+  /// [boolStrings].
+  /// For the config file, it must be a boolean.
+  bool? getOptionalBool(String key) {
+    bool? value;
+    for (final provider in _providers) {
+      value ??= provider.getOptionalBool(key);
+    }
+    return value;
+  }
+
+  /// Lookup a path in this config.
+  ///
+  /// First tries CLI argument defines, then environment variables, and
+  /// finally the config file.
+  ///
+  /// Throws if one of the configs does not contain the expected value type.
+  ///
+  /// If [resolveFileUri], resolves the paths in config file relative to the
+  /// config file.
+  ///
+  /// If [mustExist], throws if the path doesn't resolve to a file or directory
+  /// on the file system.
+  ///
+  /// Throws if one of the configs does not contain the expected value type.
+  Uri getPath(
+    String key, {
+    bool resolveFileUri = true,
+    bool mustExist = false,
+  }) {
+    final value = getOptionalPath(key,
+        resolveFileUri: resolveFileUri, mustExist: mustExist);
+    _throwIfNull(key, value);
+    return value!;
+  }
+
+  /// Lookup an optional path in this config.
+  ///
+  /// First tries CLI argument defines, then environment variables, and
+  /// finally the config file.
+  ///
+  /// Throws if one of the configs does not contain the expected value type.
+  ///
+  /// If [resolveFileUri], resolves the paths in config file relative to the
+  /// config file.
+  ///
+  /// If [mustExist], throws if the path doesn't resolve to a file or directory
+  /// on the file system.
+  Uri? getOptionalPath(
```

Comment:
These methods (and the ones on Provider) feel like they should just be an single extension on this class? Isn't this just a pretty basic wrapper around the `getOptionalString` and `getString` functions? I would expect only one implementation at least, and not each provider having to implement it :).

Reply:
After looking at the impls it looks like the main thing is they have different logic for resolving relative paths, what if instead the `Provider` class instead provided a base URI (possibly nullable) to resolve against?